Add API key validation filter

Proverbs1603 opened the pull request:
열심히 고민해서 보안을 어떻게 할 지 생각해봤는데요
스프링 Filter를 활용한 Api키 인증방식을 도입하기로 했습니다.
 1. 먼저 DB에 API_KEY 테이블을 만들어서 api_key, 만기시간, 활성여부, 클라이언트_이름 등등을 만들어두었습니다.
 2. Filter를 구현해서 /api 로 들어오는 URL에 대해 apiKey를 검증합니다. 우선은 제가 임의의 랜덤한 값을 생성해두었고 Filter로 들어온 Request의 Header에 "X-API-KEY" 를 검증하여 발급된 key인지, 만기시간이 지나지 않았는지, 활성화 되어있는지 등등을 Valid 합니다.
 3. 위를 간단히 테스트 해볼 수 있는 테스트코드를 작성하였습니다

--실패시-- 
![image](https://github.com/user-attachments/assets/eb07ba23-1cd8-48b2-a0e8-636c61ac4ccb)

<br/>

--성공시--
![image](https://github.com/user-attachments/assets/c40bfb2e-737c-4205-9841-b1188ac857ec)

Proverbs1603 committed:
Add API key validation filter

sbslc2000 commented on src/main/java/com/example/ca4u_backend/apiResponse/ApiKeyService.java:
```diff
@@ -0,0 +1,39 @@
+    return true; // 모든 조건을 만족하면 유효
+  }
+
+  private Boolean isExpired(ApiKey apiKey) {
+    LocalDateTime now = LocalDateTime.now();
+    return now.isAfter(apiKey.getExpiresAt());
+  }
+
+  private Boolean isActive(ApiKey apiKey) {
+    return apiKey.getIsActive();
+  }
```
이러한 도메인 로직은 ApiKey 객체 내부에 캡슐화하면 더 좋은 설계가 될 것 같아요.

sbslc2000 commented on src/main/java/com/example/ca4u_backend/apiResponse/ApiKeyService.java:
```diff
@@ -0,0 +1,39 @@
+    return true; // 모든 조건을 만족하면 유효
+  }
+
+  private Boolean isExpired(ApiKey apiKey) {
+    LocalDateTime now = LocalDateTime.now();
+    return now.isAfter(apiKey.getExpiresAt());
+  }
+
+  private Boolean isActive(ApiKey apiKey) {
+    return apiKey.getIsActive();
+  }
```
Boolean을 반환한 의도는 무엇인가요?

Proverbs1603 replied:
apikey가 만기되었는지 활성화되었는지 등등을 반환되는 조건문에서 확인

sbslc2000 replied:
앗 제 의도는 왜 boolean이 아닌 Boolean이냐는 의미엿습니다. Null 에 대한 고려가 필요한가요? 아래 질문도 동일합니다

Proverbs1603 replied:
아하 오.. 
혹시 boolean으로 쓰면 테이블에 들어갈 default값이 정해지나요?

Proverbs1603 replied:
위에 질문은 엔티티에 정의할 때 입니다.!

Proverbs1603 replied:
![image](https://github.com/user-attachments/assets/68d95cdc-5cf2-4f5f-a5c8-df73aeed14a1)
혹시 이런 경우에 엔티티에는 Boolean으로 정의되어있는데 메서드 return type은 아래같이 boolean으로 해도 되나요?

sbslc2000 replied:
Java boolean 기본값인 false가 매핑됩니다

sbslc2000 replied:
1. Entity와 Boolean

JPA는 Boolean이든 boolean이든 MySQL boolean 타입으로 매핑됩니다.
MYSQL boolean 타입의 실체는 tinyint로, 1비트 정수형에 해당합니다. 가질 수 있는 값은 0, 1, 그리고 null.

엔티티에 boolean으로 매핑하는 경우, 애플리케이션 레벨에서 boolean에 해당하는 값은 true 혹은 false 둘 중 하나이므로, 별도의 조작이 없다면 자연스레 DB에도 true와 false만 쌓이게 됩니다. _**별도의 값을 할당하지 않는다면 Java의 boolean 기본값인 false가 할당됩니다.**_

한편 엔티티에 Boolean으로 매핑하는 경우, 애플리케이션 레벨에서 해당 값이 null일 경우가 발생하므로 DB에도 null이 들어갈 수 있게 됩니다.

Boolean으로 매핑하는데는 이 값이 null 일 수 있을 때이어야 합니다.

2. 엔티티에는 Boolean인데 return type은 boolean으로 해도 되나요?

Boolean의 값이 null인 경우 Wrapper type이 Primitive type으로 변환되는 과정에서 에러가 발생할 수 있습니다. null이 아닌 경우에는 ok

Proverbs1603 replied:
그러면 ALTER TABLE CLUB ADD COLUMN is_active BOOLEAN NOT NULL DEFAULT true; 로 MySQL에 default 값을 true로 설정하고 엔티티와 메서드는 boolean type으로 하겠습니다

sbslc2000 replied:
대신 애플리케이션에서 ApiKey 엔티티를 생성할 때에도 true는 할당해줘야해요. 아니면 기본값인 false를 인식해서 insert 쿼리 날릴 때 false로 명시적으로 날라가서 sql 레벨의 default true가 무시될거에요ㅣ.

Proverbs1603 replied:
![image](https://github.com/user-attachments/assets/7d73479b-fbdd-4b38-baef-c23255deaa38)
@PrePersist 로 객체생성시 true로 초기화했습니다!

Proverbs1603 replied:
private boolean isActive = true; 
이게 낫나요?

sbslc2000 replied:
어떤 차이가 있을까요? 요건 퀴즈입니다

Proverbs1603 replied:
=true는 JVM 수준에서 기본값 설정이 이루어지고 PrePersist는 JPA가 객체를 저장하기 직전에 이루어진대요.
그러면 저장직전에 들어가버리니까 만약에 악성조회를 거르려고 isActive를 false로 바꾸고 나중에 JPA로 저장하면 결국 다시 true가 되어버려서 이런 경우는 안되겠네요 맞나요?

sbslc2000 commented on src/main/java/com/example/ca4u_backend/apiResponse/ApiKey.java:
```diff
@@ -0,0 +1,42 @@
+  @Column(name = "expires_at")
+  private LocalDateTime expiresAt;
+
+  @Column(name = "is_active")
+  private Boolean isActive;
```
Boolean 타입을 사용한 의도는 무엇인가요?

Proverbs1603 replied:
나중에 악성조회가 많아지면 isActive를 false로 바꿔서 apikey를 차단

sbslc2000 commented on src/main/java/com/example/ca4u_backend/filter/ApiFilter.java:
```diff
@@ -0,0 +1,60 @@
+        httpResponse.setContentType("application/json");
+        ApiResponse<Void> apiResponse = ApiResponse.fail("Invalid or missing API Key");
+        // 응답을 JSON으로 변환
+        ObjectMapper objectMapper = new ObjectMapper();
```
불필요하게 ObjectMapper 객체가 메모리에 할당되고 해제되는 일이 발생합니다. ObjectMapper instance는 스프링빈에 존재하니 주입받아 사용하면 좋을 것 같습니다.

sbslc2000 commented on src/main/java/com/example/ca4u_backend/apiResponse/ApiKey.java:
이 객체가 생성되고 사용되고 소멸하는 방식에 대해 sequence diagram 혹은 줄글로 명시해주는게 좋아보여요.

Proverbs1603 replied:
apikey를 발급하는 플랫폼은 만들 여력이 없기 때문에 수동으로 생성.
생성과 동시에 1개월 만기기간을 부여하고 나중에 해당 apikey로 조회가 들어왔을 때 만기시간이 지났다면 key를 삭제

sbslc2000 commented on src/main/java/com/example/ca4u_backend/filter/ApiFilter.java:
```diff
@@ -0,0 +1,60 @@
+      HttpServletResponse httpResponse = (HttpServletResponse) servletResponse;
+
+      // 요청에서 API 키 추출
+      String apiKey = httpRequest.getHeader("X-API-KEY");
```
X- 로 시작하는 헤더명의 역사에 대해서 스터디 해보심 좋을 것 같습니다. 엄밀히 말하면 비표준입니다.

Proverbs1603 replied:
-> API-KEY

sbslc2000 commented:
사용자는 언제 어떻게 API KEY를 획득하나요?

sbslc2000 commented on src/main/java/com/example/ca4u_backend/filter/ApiFilter.java:
```diff
@@ -0,0 +1,60 @@
+      String apiKey = httpRequest.getHeader("X-API-KEY");
+      if (apiKey == null || !apiKeyService.isValid(apiKey)) {
+        // API 키가 없거나 유효하지 않으면 요청 차단
+        httpResponse.setStatus(HttpServletResponse.SC_UNAUTHORIZED); // 401 Unauthorized
+        httpResponse.setContentType("application/json");
+        ApiResponse<Void> apiResponse = ApiResponse.fail("Invalid or missing API Key");
```
어쩌면 보안을 위해 만들어진 기능에 "너 지금 키 안보냈어"라고 친절히 알려주는 부분은 불필요하게 구현을 노출시킬지도 모릅니다. 물론 클라이언트 및 서버 단 개발하는 과정에서 디버깅이 빨라질 수 있다는 장점이 있지만, 본질적으로 어떤 정보를 보여주고 어떤 정보를 보여주지 않아야하는가는 보안적으로 고민해볼 수 있는 부분인 것 같습니다.

Proverbs1603 committed:
1. ApiKey 객체 내부에 도메인 로직 캡슐화
2. ObjectMapper 스프링 의존성 주입
3. ApiKey의 만기시간 지날시 삭제
4. X-API-KEY -> API-KEY
5. Missing key message update to "401 Unauthorized"

Proverbs1603 committed:
is_active's default value set true && type (Boolean -> boolean)

Proverbs1603 committed:
not using @prepersist but using isActive = ture
